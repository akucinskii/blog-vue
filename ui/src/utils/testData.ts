const testPosts = [
  {
    id: "313553ff-784f-497f-adfc-7e549bebab66",
    title:
      "POST TESTOWY asm.nfnasnfas.fa.sfn,acsaf;;;;;;;;;;;;jslfsafsjafjsajkfsjfsajfsafsajlfjlksajklfasjklafsjklfsajklaflkjslksfnsaf,mnfsa,fsannm",
    date: "2037-01-01",
    text: "It is a long established fact that a reader will be distracted by the readable content of a page when looking at its layout. The point of using Lorem Ipsum is that it has a more-or-less normal distribution of letters, as opposed to using 'Content here, content here', making it look like readable English. Many desktop publishing packages and web page editors now use Lorem Ipsum as their default model text, and a search for 'lorem ipsum' will uncover many web sites still in their infancy. Various versions have evolved over the years, sometimes by accident, sometimes on purpose (injected humour and the like).",
    imageUrl:
      "https://images.unsplash.com/photo-1495753379358-73c76ccd644b?ixlib=rb-1.2.1&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=1539&q=80",
    author: {
      username: "Robert",
      avatar:
        "https://images.unsplash.com/photo-1495753379358-73c76ccd644b?ixlib=rb-1.2.1&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=1539&q=80",
      description: "test description",
    },
    category: {
      id: "1",
      name: "Technology",
    },
    disabled: false,
  },
  {
    id: "313553ff-784f-497f-adfc-7e549rerab66",
    title: "sfakfasjlasdasdasndaskdjaskjsj",
    date: "2020-01-01",
    text: "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nullam euismod, nisi eu consectetur euismod, nisl nisi consectetur nisi, eu consectetur nisi nisi euismod nisi. Nullam euismod, nisi eu consectetur euismod, nisl nisi consectetur nisi, eu consectetur nisi nisi euismod nisi. Nullam euismod, nisi eu consectetur euismod, nisl nisi consectetur nisi, eu consectetur nisi nisi euismod nisi. Nullam euismod, nisi eu consectetur euismod, nisl nisi consectetur nisi, eu consectetur nisi nisi euismod nisi. Nullam euismod, nisi eu consectetur euismod, nisl nisi consectetur nisi, eu consectetur nisi nisi euismod nisi. Nullam euismod, nisi eu consectetur euismod, nisl nisi consectetur nisi, eu consectetur nisi nisi euismod nisi. Nullam euismod, nisi eu consectetur euismod, nisl nisi consectetur nisi, eu consectetur nisi nisi euismod nisi. Nullam euismod, nisi eu consectetur euismod, nisl nisi consectetur nisi, eu consectetur nisi nisi euismod nisi. Nullam euismod, nisi eu consectetur euismod, nisl nisi consectetur nisi, eu consectetur nisi nisi euismod nisi. Nullam euismod, nisi eu consectetur euismod, nisl nisi consectetur nisi, eu consectetur nisi nisi e",
    imageUrl:
      "https://images.unsplash.com/photo-1495753379358-73c76ccd644b?ixlib=rb-1.2.1&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=1539&q=80",
    author: {
      username: "Wojciech",
      avatar:
        "https://images.unsplash.com/photo-1495753379358-73c76ccd644b?ixlib=rb-1.2.1&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=1539&q=80",
      description: "test description",
    },
    category: {
      id: "2",
      name: "Dieta",
    },
    disabled: false,
  },
  {
    id: "fsafasd",
    title: "Najlepsza dieta w okolicy. Zobacz teraz!!!",
    date: "2020-01-01",
    text: "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nullam euismod, nisi eu consectetur euismod, nisl nisi consectetur nisi, eu consectetur nisi nisi euismod nisi. Nullam euismod, nisi eu consectetur euismod, nisl nisi consectetur nisi, eu consectetur nisi nisi euismod nisi. Nullam euismod, nisi eu consectetur euismod, nisl nisi consectetur nisi, eu consectetur nisi nisi euismod nisi. Nullam euismod, nisi eu consectetur euismod, nisl nisi consectetur nisi, eu consectetur nisi nisi euismod nisi. Nullam euismod, nisi eu consectetur euismod, nisl nisi consectetur nisi, eu consectetur nisi nisi euismod nisi. Nullam euismod, nisi eu consectetur euismod, nisl nisi consectetur nisi, eu consectetur nisi nisi euismod nisi. Nullam euismod, nisi eu consectetur euismod, nisl nisi consectetur nisi, eu consectetur nisi nisi euismod nisi. Nullam euismod, nisi eu consectetur euismod, nisl nisi consectetur nisi, eu consectetur nisi nisi euismod nisi. Nullam euismod, nisi eu consectetur euismod, nisl nisi consectetur nisi, eu consectetur nisi nisi euismod nisi. Nullam euismod, nisi eu consectetur euismod, nisl nisi consectetur nisi, eu consectetur nisi nisi e",
    imageUrl:
      "https://images.unsplash.com/photo-1529565214304-a882ebc5a8e6?ixlib=rb-1.2.1&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=2070&q=80",
    author: {
      username: "Test author",
      avatar:
        "https://images.unsplash.com/photo-1633621412960-6df85eff8c85?ixlib=rb-1.2.1&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=627&q=80",
      description: "test description",
    },
    category: {
      id: "1",
      name: "Technology",
    },
    disabled: false,
  },
  {
    id: "fsafasd",
    title: "This post shouldnt be visible",
    date: "2020-01-01",
    text: "This post shouldnt be visible",
    imageUrl:
      "https://images.unsplash.com/photo-1529565214304-a882ebc5a8e6?ixlib=rb-1.2.1&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=2070&q=80",
    author: {
      username: "Test author",
      avatar:
        "https://images.unsplash.com/photo-1633621412960-6df85eff8c85?ixlib=rb-1.2.1&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=627&q=80",
      description: "test description",
    },
    category: {
      id: "2",
      name: "Dieta",
    },
    disabled: true,
  },
];

export const testCategories = [
  {
    id: "1",
    name: "Technology",
  },
  {
    id: "2",
    name: "Dieta",
  },
  {
    id: "3",
    name: "Entertainment",
  },
];

export const testUser = {
  id: "1",
  username: "TestUser",
  password: "Password123.",
  avatar_url: "testAvatar.jpg",
  description: "test description",
  name: "Test name",
  surname: "Test surname",
  is_admin: true,
};

export default testPosts;
