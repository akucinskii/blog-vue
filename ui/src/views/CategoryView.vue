<script lang="ts" setup>
import PostList from "../components/Posts/PostList.vue";
import ArticleThumbnail from "../components/Posts/Article/ArticleThumbnail.vue";
import { ref, watch } from "vue";
import { useRoute } from "vue-router";
import { usePostsStore } from "../stores/posts";
import { useCategoriesStore } from "../stores/categories";
import { ArticleInterface, CategoryInterface } from "../types/types";
import { computed } from "vue";

const route = useRoute();
const postStore = usePostsStore();
const categoriesStore = useCategoriesStore();
const category = ref<undefined | CategoryInterface>({} as CategoryInterface);
const posts = ref([] as ArticleInterface[]);

const fetchPostsByCategory = async () => {
  await postStore.fetchPosts();
  await categoriesStore.fetchCategories();

  const uuid = Array.isArray(route.params.id) ? route.params.id[0] : route.params.id;

  category.value = categoriesStore.getCategoryByUUID(uuid);
  posts.value = postStore.getPostsByCategory(uuid);
};

const data = computed(() => {
  return posts.value.filter((post) => post.disabled === false);
});

fetchPostsByCategory();

watch(route, fetchPostsByCategory);

defineExpose({
  route,
  posts,
  category,
});
</script>

<template>
  <v-main class="mt-8">
    <v-col v-if="category" id="categoryTitle" class="text-center"
      ><h1>Posts in {{ category.name }}</h1>
    </v-col>
    <PostList v-if="data.length > 0">
      <v-col v-for="post in data" :key="post.id" class="post" cols="12" md="6" xl="6">
        <ArticleThumbnail
          v-if="!post.disabled"
          :id="post.id"
          :url="post.image_url"
          :author="post.author"
          :title="post.title"
          :category="post.category.name"
          :text="post.text"
        />
      </v-col>
    </PostList>
    <v-col v-else class="text-center"><h3>No posts found in this category :(</h3></v-col>
  </v-main>
</template>

<script lang="scss" scoped></script>
