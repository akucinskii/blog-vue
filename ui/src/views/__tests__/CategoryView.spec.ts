/* eslint-disable @typescript-eslint/no-empty-function */
import CategoryView from "../CategoryView.vue";
import { mount } from "@vue/test-utils";
import { useRoute } from "vue-router";
import { createTestingPinia } from "@pinia/testing";
import testData, { testCategories } from "../../utils/testData";
import { nextTick } from "vue";

jest.mock("vue-router", () => ({
  useRoute: jest.fn(),
}));

describe("Category page", () => {
  test("renders correct error message", async () => {
    const NoPostFoundMessage = "No posts found in this category :(";

    (useRoute as jest.Mock).mockImplementationOnce(() => ({
      params: {
        id: "IDLikeThisDoesntExist",
      },
    }));

    const wrapper = mount(CategoryView, {
      global: {
        plugins: [
          createTestingPinia({
            initialState: {
              posts: {
                posts: testData,
              },
              categories: {
                categories: testCategories,
              },
            },
          }),
        ],
        stubs: ["router-link", "router-view"],
      },
    });

    await nextTick();

    expect(wrapper.find("h3").text()).toContain(NoPostFoundMessage);
  });

  test("renders correct posts when id is given as string", async () => {
    // Test category name of "technology" has id of 1
    const categoryId = "1";

    (useRoute as jest.Mock).mockImplementationOnce(() => ({
      params: {
        id: categoryId,
      },
    }));

    const wrapper = mount(CategoryView, {
      global: {
        plugins: [
          createTestingPinia({
            initialState: {
              posts: {
                posts: testData,
              },
              categories: {
                categories: testCategories,
              },
            },
          }),
        ],
        stubs: ["router-link", "router-view"],
      },
    });

    await nextTick();
    await nextTick();
    await nextTick();

    // Test data has 2 posts with category with name "Technology"
    expect(wrapper.findAll(".post").length).toBe(2);
  });

  test("Renders correct posts when CategoryId is given as array", async () => {
    const categoryIdArray = ["1", "2"];

    (useRoute as jest.Mock).mockImplementationOnce(() => ({
      params: {
        id: categoryIdArray,
      },
    }));

    const wrapper = mount(CategoryView, {
      global: {
        plugins: [
          createTestingPinia({
            initialState: {
              posts: {
                posts: testData,
              },
              categories: {
                categories: testCategories,
              },
            },
          }),
        ],
        stubs: ["router-link", "router-view"],
      },
    });

    await nextTick();
    await nextTick();
    await nextTick();

    expect(wrapper.findAll(".post").length).toBe(2);
  });

  test("Renders correct header", async () => {
    const text = "Posts in Technology";
    const categoryId = "1";

    (useRoute as jest.Mock).mockImplementationOnce(() => ({
      params: {
        id: categoryId,
      },
    }));

    const wrapper = mount(CategoryView, {
      global: {
        plugins: [
          createTestingPinia({
            initialState: {
              posts: {
                posts: testData,
              },
              categories: {
                categories: testCategories,
              },
            },
          }),
        ],
        stubs: ["router-link", "router-view"],
      },
    });

    await nextTick();
    await nextTick();

    expect(wrapper.find("h1").text()).toMatch(/Posts\sin\s\w+/g);
    expect(wrapper.find("h1").text()).toEqual(text);
  });

  test("Should match snapshot", async () => {
    const categoryId = "1";

    (useRoute as jest.Mock).mockImplementationOnce(() => ({
      params: {
        id: categoryId,
      },
    }));

    const wrapper = mount(CategoryView, {
      global: {
        plugins: [
          createTestingPinia({
            initialState: {
              posts: {
                posts: testData,
              },
              categories: {
                categories: testCategories,
              },
            },
          }),
        ],
        stubs: ["router-link", "router-view"],
      },
    });

    await nextTick();

    expect(wrapper.element).toMatchSnapshot();
  });
});
