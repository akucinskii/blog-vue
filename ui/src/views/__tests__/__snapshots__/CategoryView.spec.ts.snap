// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Category page Should match snapshot 1`] = `
<v-main
  class="mt-8"
>
  <v-col
    class="text-center"
    id="categoryTitle"
  >
    <h1>
      Posts in 
    </h1>
  </v-col>
  <v-col
    class="text-center"
  >
    <h3>
      No posts found in this category :(
    </h3>
  </v-col>
</v-main>
`;
